### PR TITLE
fix ser/de 

### PR DESCRIPTION
**Pull request by LesnyRumcajs**
Went into the nim code to find the true definitions for services + corrected invalid serialization and deserialization.

Now we need to get the RLN proof

```
Got swarm event Some(Behaviour(LightPush(Message { peer: PeerId("16Uiu2HAkvjxLmQEV1bKerZh9FRXoVERhMDuyMyWQu251Bc4okCph"), message: Response { request_id: OutboundRequestId(1), response: PushRpc { request_id: "0", request: None, response: Some(PushResponse { is_success: false, info: Some("RLN validation failed") }) } } })))
```

### CODE REVIEW

**Comment by LesnyRumcajs** [comment]
Peer exchange now also returns non-empty responses (to be formatted properly in logs)
```
     Running `target/debug/waku-oxidized --peers /ip4/164.90.173.33/tcp/30304 /light-cthulhu/1/message/proto kraaa`
[2024-05-25T13:11:56Z INFO  waku_oxidized] Libp2p local peer id: PeerId("12D3KooWSadeQMhBPBMTbphY3mzCFFomhb5ATivgbn3Vg3J41iZ8")
[2024-05-25T13:11:56Z DEBUG multistream_select::dialer_select] Dialer: Proposed protocol: /noise
[2024-05-25T13:11:56Z DEBUG multistream_select::dialer_select] Dialer: Expecting proposed protocol: /noise
[2024-05-25T13:11:56Z DEBUG multistream_select::negotiated] Negotiated: Received confirmation for protocol: /noise
[2024-05-25T13:11:56Z DEBUG multistream_select::dialer_select] Dialer: Proposed protocol: /yamux/1.0.0
[2024-05-25T13:11:56Z DEBUG multistream_select::dialer_select] Dialer: Expecting proposed protocol: /yamux/1.0.0
[2024-05-25T13:11:56Z DEBUG yamux::connection] new connection: 384c6160 (Client)
[2024-05-25T13:11:56Z DEBUG yamux::connection::rtt] sending ping 2514450403
Connection estabilished with PeerId("16Uiu2HAkvjxLmQEV1bKerZh9FRXoVERhMDuyMyWQu251Bc4okCph") on Dialer { address: "/ip4/164.90.173.33/tcp/30304", role_override: Dialer }
[2024-05-25T13:11:56Z DEBUG yamux::connection] 384c6160: new outbound (Stream 384c6160/1) of (Connection 384c6160 Client (streams 0))
[2024-05-25T13:11:56Z DEBUG multistream_select::dialer_select] Dialer: Proposed protocol: /vac/waku/lightpush/2.0.0-beta1
[2024-05-25T13:11:56Z DEBUG yamux::connection] 384c6160: new outbound (Stream 384c6160/3) of (Connection 384c6160 Client (streams 1))
[2024-05-25T13:11:56Z DEBUG multistream_select::dialer_select] Dialer: Proposed protocol: /vac/waku/peer-exchange/2.0.0-alpha1
[2024-05-25T13:11:56Z DEBUG multistream_select::negotiated] Negotiated: Received confirmation for protocol: /yamux/1.0.0
[2024-05-25T13:11:56Z DEBUG yamux::connection::rtt] received pong 2514450403, estimated round-trip-time 20.283541ms
[2024-05-25T13:11:56Z DEBUG multistream_select::listener_select] Listener: confirming protocol: /vac/waku/metadata/1.0.0
[2024-05-25T13:11:56Z DEBUG multistream_select::listener_select] Listener: sent confirmed protocol: /vac/waku/metadata/1.0.0
[2024-05-25T13:11:56Z DEBUG multistream_select::listener_select] Listener: rejecting protocol: /vac/waku/relay/2.0.0
[2024-05-25T13:11:56Z DEBUG multistream_select::dialer_select] Dialer: Received confirmation for protocol: /vac/waku/lightpush/2.0.0-beta1
[2024-05-25T13:11:56Z DEBUG multistream_select::dialer_select] Dialer: Received confirmation for protocol: /vac/waku/peer-exchange/2.0.0-alpha1
[2024-05-25T13:11:56Z DEBUG multistream_select::listener_select] Listener: rejecting protocol: /ipfs/id/1.0.0
[2024-05-25T13:11:56Z DEBUG multistream_select::listener_select] Listener: rejecting protocol: /meshsub/1.1.0
[2024-05-25T13:11:56Z DEBUG yamux::connection::stream] 384c6160/3: eof
Got swarm event Some(Behaviour(PeerExchange(Message { peer: PeerId("16Uiu2HAkvjxLmQEV1bKerZh9FRXoVERhMDuyMyWQu251Bc4okCph"), message: Response { request_id: OutboundRequestId(1), response: PeerExchangeRpc { query: Some(PeerExchangeQuery { num_peers: 0 }), response: Some(PeerExchangeResponse { peer_infos: [PeerInfo { enr: [248, 181, 184, 64, 84, 230, 170, 173, 81, 27, 173, 228, 88, 156, 195, 143, 16, 167, 203, 151, 96, 106, 227, 156, 172, 87, 151, 81, 37, 95, 101, 185, 202, 55, 179, 176, 9, 239, 147, 91, 34, 243, 191, 44, 163, 9, 21, 151, 194, 23, 221, 74, 69, 228, 172, 117, 56, 30, 170, 233, 122, 2, 157, 45, 136, 49, 156, 219, 1, 130, 105, 100, 130, 118, 52, 130, 105, 112, 132, 207, 180, 223, 67, 138, 109, 117, 108, 116, 105, 97, 100, 100, 114, 115, 128, 130, 114, 115, 147, 0, 1, 8, 0, 0, 0, 1, 0, 2, 0, 3, 0, 4, 0, 5, 0, 6, 0, 7, 137, 115, 101, 99, 112, 50, 53, 54, 107, 49, 161, 3, 254, 216, 179, 228, 128, 160, 108, 77, 89, 78, 209, 235, 223, 16, 64, 0, 70, 25, 190, 239, 44, 246, 66, 184, 76, 213, 64, 178, 11, 110, 170, 80, 131, 116, 99, 112, 130, 118, 96, 131, 117, 100, 112, 130, 35, 45, 133, 119, 97, 107, 117, 50, 15] }, PeerInfo { enr: [248, 181, 184, 64, 128, 32, 201, 131, 60, 81, 84, 50, 203, 125, 127, 30, 236, 57, 10, 54, 62, 96, 193, 250, 204, 105, 121, 204, 180, 104, 178, 198, 217, 235, 25, 173, 32, 113, 228, 57, 246, 37, 127, 185, 95, 168, 133, 17, 198, 124, 40, 238, 160, 134, 212, 145, 139, 177, 217, 83, 149, 48, 177, 228, 20, 166, 188, 153, 1, 130, 105, 100, 130, 118, 52, 130, 105, 112, 132, 213, 136, 92, 165, 138, 109, 117, 108, 116, 105, 97, 100, 100, 114, 115, 128, 130, 114, 115, 147, 0, 1, 8, 0, 0, 0, 1, 0, 2, 0, 3, 0, 4, 0, 5, 0, 6, 0, 7, 137, 115, 101, 99, 112, 50, 53, 54, 107, 49, 161, 3, 152, 39, 203, 115, 55, 146, 72, 48, 37, 185, 236, 249, 222, 242, 254, 153, 90, 221, 14, 199, 101, 229, 83, 36, 53, 252, 247, 74, 8, 179, 40, 161, 131, 116, 99, 112, 130, 118, 96, 131, 117, 100, 112, 130, 35, 45, 133, 119, 97, 107, 117, 50, 15] }, PeerInfo { enr: [248, 181, 184, 64, 181, 20, 79, 127, 129, 150, 195, 224, 43, 158, 93, 3, 93, 172, 54, 201, 3, 4, 130, 100, 213, 48, 21, 22, 26, 116, 162, 176, 59, 87, 29, 73, 126, 111, 216, 249, 38, 249, 52, 94, 88, 88, 63, 234, 41, 149, 176, 19, 205, 13, 75, 192, 90, 212, 70, 138, 4, 4, 147, 64, 42, 86, 51, 193, 1, 130, 105, 100, 130, 118, 52, 130, 105, 112, 132, 5, 45, 97, 204, 138, 109, 117, 108, 116, 105, 97, 100, 100, 114, 115, 128, 130, 114, 115, 147, 0, 1, 8, 0, 0, 0, 1, 0, 2, 0, 3, 0, 4, 0, 5, 0, 6, 0, 7, 137, 115, 101, 99, 112, 50, 53, 54, 107, 49, 161, 3, 242, 173, 0, 97, 121, 131, 152, 167, 163, 248, 26, 13, 144, 45, 71, 148, 223, 152, 141, 246, 206, 59, 6, 126, 94, 97, 181, 142, 195, 62, 151, 245, 131, 116, 99, 112, 130, 118, 96, 131, 117, 100, 112, 130, 35, 45, 133, 119, 97, 107, 117, 50, 15] }, PeerInfo { enr: [248, 181, 184, 64, 124, 96, 178, 248, 71, 203, 65, 184, 199, 164, 166, 164, 56, 31, 147, 206, 23, 192, 167, 121, 194, 121, 225, 112, 35, 145, 129, 97, 16, 111, 148, 34, 6, 177, 55, 96, 27, 117, 225, 195, 152, 93, 51, 153, 152, 179, 116, 26, 116, 252, 214, 100, 88, 173, 218, 214, 175, 124, 78, 74, 105, 96, 2, 237, 1, 130, 105, 100, 130, 118, 52, 130, 105, 112, 132, 80, 71, 227, 151, 138, 109, 117, 108, 116, 105, 97, 100, 100, 114, 115, 128, 130, 114, 115, 147, 0, 1, 8, 0, 0, 0, 1, 0, 2, 0, 3, 0, 4, 0, 5, 0, 6, 0, 7, 137, 115, 101, 99, 112, 50, 53, 54, 107, 49, 161, 2, 10, 104, 191, 127, 28, 68, 171, 13, 45, 44, 74, 34, 14, 199, 59, 36, 132, 29, 53, 32, 167, 194, 96, 123, 88, 40, 170, 62, 11, 160, 59, 208, 131, 116, 99, 112, 130, 118, 96, 131, 117, 100, 112, 130, 35, 45, 133, 119, 97, 107, 117, 50, 15] }, PeerInfo { enr: [248, 181, 184, 64, 31, 73, 134, 177, 72, 198, 195, 42, 204, 245, 154, 235, 49, 48, 236, 81, 27, 224, 47, 24, 139, 20, 22, 120, 103, 103, 232, 162, 126, 34, 0, 103, 41, 30, 80, 81, 185, 93, 3, 244, 18, 231, 53, 100, 165, 232, 96, 230, 210, 86, 154, 36, 220, 191, 184, 71, 2, 59, 131, 27, 100, 252, 1, 20, 1, 130, 105, 100, 130, 118, 52, 130, 105, 112, 132, 144, 91, 107, 49, 138, 109, 117, 108, 116, 105, 97, 100, 100, 114, 115, 128, 130, 114, 115, 147, 0, 1, 8, 0, 0, 0, 1, 0, 2, 0, 3, 0, 4, 0, 5, 0, 6, 0, 7, 137, 115, 101, 99, 112, 50, 53, 54, 107, 49, 161, 3, 100, 160, 22, 236, 201, 33, 132, 111, 227, 67, 157, 92, 180, 19, 93, 44, 236, 95, 52, 175, 220, 202, 192, 96, 88, 207, 136, 199, 202, 76, 9, 160, 131, 116, 99, 112, 130, 118, 96, 131, 117, 100, 112, 130, 35, 45, 133, 119, 97, 107, 117, 50, 15] }] }) } } })))
[2024-05-25T13:11:56Z DEBUG yamux::connection::stream] 384c6160/1: eof
[2024-05-25T13:11:56Z DEBUG multistream_select::listener_select] Listener: rejecting protocol: /meshsub/1.0.0
Got swarm event Some(Behaviour(LightPush(Message { peer: PeerId("16Uiu2HAkvjxLmQEV1bKerZh9FRXoVERhMDuyMyWQu251Bc4okCph"), message: Response { request_id: OutboundRequestId(1), response: PushRpc { request_id: "0", request: None, response: Some(PushResponse { is_success: false, info: Some("RLN validation failed") }) } } })))
```

File: src/lib.rs [hunk 1]
```diff
@@ -65,25 +65,40 @@ impl WakuLightNode {
...
-            light_push::messages::PushRequest {
-                pubsub_topic: DEFAULT_PUBSUB_TOPIC.to_string(),
-                message: Some(light_push::message::WakuMessage {
-                    content_topic,
-                    payload,
-                    ephemeral: Some(false),
-                    meta: None,
-                    timestamp: None,
-                    version: None,
+            light_push::messages::PushRpc {
```
Comment (ameba23) [inline src/lib.rs:87]: Ah, so this is what we were doing wrong.  Strange that request and response are optional.  But now i think about it, with protobuf v3, all fields are optional - so im not sure why people sometimes anyway annotate them as optional.  Great that you figured this out!

File: src/light_push.rs [hunk 2]
```diff
@@ -24,37 +21,31 @@ pub struct Codec {}
...
         T: AsyncRead + Unpin + Send,
     {
-        let mut vec = Vec::new();
-
-        io.take(REQUEST_SIZE_MAXIMUM).read_to_end(&mut vec).await?;
-
-        let request = messages::PushRequest::decode(&vec[..])?;
+        let mut vec = Vec::with_capacity(MAX_LIGHTPUSH_RPC_SIZE as usize);
+        io.read_to_end(&mut vec).await?;
+        let request = Self::Request::decode_length_delimited(&vec[..])?;
```
Comment (ameba23) [inline src/light_push.rs:33]: was using `decode_length_delimited` needed to get it to work, or is it just a better way of doing it?

**Review by ameba23 — APPROVED** [review]
amazing work :fire:

ok so now we need to do these proofs and i have no idea how they work.  I guess need to do a bit of reading.  I was really hoping we wouldn't need to. Last time i played around with waku they were still optional.

**Pull request merged**